tunnel-server/src/auth.ts

```diff
@@ -1,55 +1,123 @@
+  role: z.string(),
+  type: z.string(),
+  scopes: z.array(z.string()),
+  sub: z.string(),
+  exp: z.number().optional()
+})
+
+export type ClaimsSchema = typeof claimsSchema
+
+export type Claims = z.infer<ClaimsSchema>
+
+export type AuthorizationHeader = BasicAuthorizationHeader | BearerAuthorizationHeader
+
+export type BasicAuthorizationHeader = {
+  scheme: 'Basic'
+  username: string
+  password: string
+}
+
+export type BearerAuthorizationHeader = {
+  scheme: 'Bearer'
+  token: string
 }
 
-function extractBasicAuth(req: IncomingMessage){
+const cookieName = 'preevy-sass-jwt'
+
+function extractAuthorizationHeader(req: IncomingMessage): AuthorizationHeader | undefined {
   const authorization = req.headers['authorization'];
   const [scheme, data] = authorization?.split(' ') ?? []
-  if (scheme !== 'Basic') {
-    return undefined
+  if (scheme === 'Bearer') {
+    return { scheme, token: data }
   }
-  const basicAuth = Buffer.from(data, 'base64').toString('ascii');
-  const sep = basicAuth.indexOf(":")
-  if (sep === -1){
-    return undefined
+  if (scheme === 'Basic') {
+    const basicAuth = Buffer.from(data, 'base64').toString('ascii');
+    const sep = basicAuth.indexOf(":")
+    if (sep === -1){
+      return undefined
+    }
+    return {scheme, username: basicAuth.slice(0,sep), password: basicAuth.slice(sep+1)};
   }
-  return [basicAuth.slice(0,sep), basicAuth.slice(sep+1)];
+  return undefined
 }
 
-export function tunnelingKeyAuthenticator(publicKey: KeyObject){
-  return async (req: IncomingMessage)=>{
-    const auth = extractBasicAuth(req);
-    if (!auth){
-      return;
-    }
-    const [username, rawJWT] = auth;
-    if (username !== 'x-preevy-profile-key'){
-      return;
+export function JwtAuthenticator(env: PreviewEnv){
+  return async (req: IncomingMessage):Promise<AuthenticationResult>=>{
+    const auth = extractAuthorizationHeader(req);
+    const jwt = match(auth)
+      .with({scheme: 'Basic', username: 'x-preevy-profile-key'}, ({password:jwt}) => jwt)
+      .with({scheme: 'Bearer'}, ({token:jwt}) => jwt)
+      .otherwise(() => new Cookies(req, undefined as any).get(cookieName))
+
+    if (!jwt){
+      return {isAuthenticated: false}
     }
-    const thumbprint = await calculateJwkThumbprintUri(await exportJWK(publicKey))
-    const token = await jwtVerify(rawJWT, publicKey, {issuer: `preevy://${thumbprint}`})
+
+    const { iss } = decodeJwt(jwt)
+    const { issuer, publicKey, extractClaims } = await match(iss).when(
+     ()=>iss?.startsWith('preevy://'), async () => {
+        const thumbprint = await calculateJwkThumbprintUri(await exportJWK(env.publicKey))
+        return {
+          publicKey: env.publicKey,
+          issuer: `preevy://${thumbprint}`,
+          extractClaims: (token:JWTPayload)=>({
+            role: 'admin',
+            type: 'profile',
+            exp: token.exp,
+            scopes: ['admin'],
+            sub: `preevy-profile:${env.publicKeyThumbprint}`
```

Comment:
this is not exactly the profile but rather the public key thumbprint
```suggestion
            sub: `preevy-pubkey:${env.publicKeyThumbprint}`
```